merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java

```diff
@@ -728,19 +748,50 @@ public static SimulationResults computeResults(
       topics.add(Triple.of(topics.size(), serializableTopic.name(), serializableTopic.outputType().getSchema()));
     }
 
-    final var serializedTimeline = new TreeMap<Duration, List<EventGraph<Pair<Integer, SerializedValue>>>>();
+    final var spanToActivities = spanToSimulatedActivities(engine,spanInfo);
+    final var serializedTimeline = new TreeMap<Duration, List<EventGraph<EventRecord>>>();
     var time = Duration.ZERO;
     for (var point : timeline.points()) {
       if (point instanceof TemporalEventSource.TimePoint.Delta delta) {
         time = time.plus(delta.delta());
       } else if (point instanceof TemporalEventSource.TimePoint.Commit commit) {
         final var serializedEventGraph = commit.events().substitute(
             event -> {
-              EventGraph<Pair<Integer, SerializedValue>> output = EventGraph.empty();
+              // TODO can we do this more efficiently?
+              EventGraph<EventRecord> output = EventGraph.empty();
               for (final var serializableTopic : serializableTopics) {
                 Optional<SerializedValue> serializedEvent = trySerializeEvent(event, serializableTopic);
                 if (serializedEvent.isPresent()) {
-                  output = EventGraph.concurrently(output, EventGraph.atom(Pair.of(serializableTopicToId.get(serializableTopic), serializedEvent.get())));
+
+
+                  // Make sure we have a true mapping between span and events
+                  // if not loop back up the parents untill we find one
+                  // we can potentially not find a parent
+                  var activitySpanID = Optional.ofNullable(spanToActivities.get(event.provenance()).id());
+                  if (activitySpanID.isEmpty()) {
```

Comment:
I'm going to need to pore over this section for a little longer - leaving a comment for now so I can submit the rest of this review

Reply:
Experimenting here... not sure this quite has all the properties we want yet, but don't want to leave you hanging too long

```java
// If the event's `provenance` has no simulated activity id, search its ancestors to find the nearest
// simulated activity id, if one exists
if (!spanToActivities.containsKey(event.provenance())) {
  var spanId = event.provenance();
  while (true) {
    if (spanToActivities.containsKey(spanId)) {
      spanToActivities.put(event.provenance(), spanToActivities.get(spanId));
      break;
    }
    final var parentSpanId = engine.getSpan(spanId).parent();
    if (parentSpanId.isEmpty()) {
      break;
    }
  }
```

Reply:
Modified the above code a little and added it.

Reply:
Looks good to me! Good catch that it was never updating `spanId` 🤦